pkg/util/cache/cache.go

```diff
@@ -81,100 +187,94 @@ func (c *lruCache[K, V]) GetAndPin(key K) (*cacheItem[K, V], bool) {
 		item.pinCount.Inc()
 
 		c.rwlock.Unlock()
-		return item, true
+		return item, nil
 	}
 
 	c.rwlock.Unlock()
-	if c.onCacheMiss != nil {
-		value, ok := c.onCacheMiss(key)
+	if c.loader != nil {
+		// Try scavenge if there is room. If not, fail fast.
+		//	Note that the test is not accurate since we are not holding the lock here.
+		if _, ok := c.tryScavenge(key); !ok {
+			return nil, ErrNotEnoughSpace
+		}
+		value, ok := c.loader(key)
 		if ok {
-			c.rwlock.Lock()
-			defer c.rwlock.Unlock()
-			item := c.setAndPin(key, value)
-			return item, true
+			item, err := c.setAndPin(key, value)
+			if err != nil {
+				return nil, err
+			}
+			return item, nil
 		}
 	}
 
-	return nil, false
+	return nil, ErrNoSuchItem
 }
 
-func (c *lruCache[K, V]) Contain(key K) bool {
-	c.rwlock.RLock()
-	defer c.rwlock.RUnlock()
-
-	_, ok := c.items[key]
-	return ok
+func (c *lruCache[K, V]) tryScavenge(key K) ([]K, bool) {
+	ok, collector := c.scavenger.Collect(key)
+	toEvict := make([]K, 0)
+	if !ok {
+		// Try evict to make room
+		done := false
+		for p := c.accessList.Back(); p != nil && !done; p = p.Prev() {
```

Comment:
I am not sure whether `accessList` is thread(goroutine) safe.
If we have a goroutine that is pushing item to `accessList` and another is iterating it, maybe there's data race problem.